clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -191,45 +194,82 @@ public void updateFetchPositions(Set<TopicPartition> partitions) {
         if (topics != null && topics.isEmpty())
             return Collections.emptyMap();
 
-        final HashMap<String, List<PartitionInfo>> topicsPartitionInfos = new HashMap<>();
-        long startTime = time.milliseconds();
+        long start = time.milliseconds();
+        long remaining = timeout;
 
-        while (time.milliseconds() - startTime < timeout) {
-            RequestFuture<ClientResponse> requestFuture = sendMetadataRequest(topics);
-            if (requestFuture != null) {
-                client.poll(requestFuture);
+        do {
+            RequestFuture<ClientResponse> future = sendMetadataRequest(topics);
+            client.poll(future, remaining);
 
-                if (requestFuture.succeeded()) {
-                    MetadataResponse response =
-                        new MetadataResponse(requestFuture.value().responseBody());
+            if (future.failed() && !future.isRetriable())
+                throw future.exception();
 
-                    for (String topic : response.cluster().topics())
-                        topicsPartitionInfos.put(
-                            topic, response.cluster().availablePartitionsForTopic(topic));
+            if (future.succeeded() && !future.value().wasDisconnected()) {
+                MetadataResponse response = new MetadataResponse(future.value().responseBody());
+                Cluster cluster = response.cluster();
+
+                Set<String> unauthorizedTopics = cluster.unauthorizedTopics();
+                if (!unauthorizedTopics.isEmpty())
+                    throw new TopicAuthorizationException(unauthorizedTopics);
```

Comment:
Actually the topic metadata request will only return authorized topics when you send an empty topic list. So the only time this will be thrown is when the user explicitly requests a topic that they are not authorized to view.